content/blog/2020-07-03-optimizing-dvc-remotes.md

```diff
@@ -0,0 +1,390 @@
+was to improve performance in this area.
+
+![](/uploads/images/2020-07-03/optimization_screenshot.png '=600')
+
+In particular, we identified one bottleneck which affects all cloud sync tools -
+remote status queries. Any data sync tool will be restricted by certain limits
+imposed by cloud storage APIs - namely, the actual API methods available for
+querying file status. Using one method over another could mean the difference
+between a given sync operation taking seconds rather than hours to complete.
+
+And with the release of version 1.0, DVC now includes several performance
+enhancements to address this issue, ensuring that we always optimize requests
+made to cloud storage APIs.
+
+_Note: "Cloud storage" and "remote storage" will be used interchangeably
+throughout this post. When discussing dataset size in this post, we mean size in
+terms of total number of files in a dataset, rather than the total amount of
+file data (bytes)._
+
+## Why status queries are a performance bottleneck
+
+Using cloud storage will require some solution for synchronizing data between
+the cloud and a user's local machine. Many general-use tools are available for
+synchronizing data to and from cloud storage, some widely used options are
+[rsync](https://rsync.samba.org/), [rclone](https://rclone.org/) and
+[aws sync](https://docs.aws.amazon.com/cli/latest/reference/s3/sync.html), each
+with their own advantages and disadvantages. Likewise, in DVC we provide sync
+functionality through our [remote](/doc/command-reference/remote) commands.
+
+Regardless of use case, any data sync tool must solve one particular issue:
+
+**Determining which files to upload or download during a sync operation.**
+
+In order to sync data between a local machine and remote storage, we must do the
+following:
+
+1. Determine which files are present locally.
+2. Query the cloud storage API to determine which files are present in the
+   cloud.
+3. Compute the difference between the two sets of files.
+
+Once this file status has been determined, the necessary files can then be
+copied to or from cloud storage. Files which only exist locally, must be
+uploaded. Likewise, files which only exist in cloud storage must be downloaded.
+
+While this may seem like a simple problem, it actually represents a significant
+performance bottleneck for data synchronization, particularly becuse of step #2.
+When querying a cloud storage API, a data sync tool will be restricted by any
+limits imposed by the API itself.
+
+However, with the optimizations new to DVC 1.0, we are now able to work around
+these limitations and provide substantial performance benefits over older DVC
+releases and general data sync tools. In fact, DVC 1.0 offers improved runtimes
+over rclone by 20x or more in certain scenarios.
+
+### DVC 1.0 vs DVC 0.91 vs rclone performance
+
+![benchmarks](/uploads/images/2020-07-03/dvc_rclone_bench.svg 'DVC 1.0 vs rclone performance comparison')
```

Comment:
this is great improvement! 

a few comments though - we don't probably care about read time _and_ user time comparison, real time is enough. Instead, can we put other scenarios here - one file to 1M, 1M to empty, some other interesting cases? like rclone with no traverse? it's better to name _use case_ and show total time per use case